workflow/controller/hooks.go

```diff
@@ -0,0 +1,92 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+	"github.com/argoproj/argo-workflows/v3/util/expr/argoexpr"
+	"github.com/argoproj/argo-workflows/v3/util/expr/env"
+	"github.com/argoproj/argo-workflows/v3/workflow/common"
+	"github.com/argoproj/argo-workflows/v3/workflow/templateresolution"
+)
+
+func IsExitHook(hookName wfv1.LifecycleEvent) bool {
+	return hookName == wfv1.ExitLifecycleEvent
+}
+
+func (woc *wfOperationCtx) executeWfLifeCycleHook(ctx context.Context, tmplCtx *templateresolution.Context) error {
+	for hookName, hook := range woc.wf.Spec.Hooks {
+		//exit hook will be execute in runOnExitNode
+		if IsExitHook(hookName) {
+			continue
+		}
+		execute, err := shouldExecuteHook(hook.Expression, woc.globalParams)
+		if err != nil {
+			return err
+		}
+		if execute {
+			hookNodeName := generateLifeHookNodeName(woc.wf.ObjectMeta.Name, string(hookName))
+			woc.log.WithField("lifeCycleHook", hookName).WithField("node", hookNodeName).Infof("Running workflow level hooks")
+			_, err := woc.executeTemplate(ctx, hookNodeName, &wfv1.WorkflowStep{Template: hook.Template}, tmplCtx, hook.Arguments, &executeTemplateOpts{})
+			if err != nil {
+				return err
+			}
+			woc.addChildNode(woc.wf.Name, hookNodeName)
+		}
+	}
+
+	return nil
+}
+
+func (woc *wfOperationCtx) executeTmplLifeCycleHook(ctx context.Context, scope *wfScope, lifeCycleHooks wfv1.LifecycleHooks, parentNode *wfv1.NodeStatus, boundaryID string, tmplCtx *templateresolution.Context, prefix string) error {
+
+	for hookName, hook := range lifeCycleHooks {
+		//exit hook will be execute in runOnExitNode
+		if IsExitHook(hookName) {
+			continue
+		}
+		execute, err := shouldExecuteHook(hook.Expression, woc.globalParams.Merge(scope.getParameters()))
```

Comment:
I'm not a fan of passing parameters and then doing some sed on the names. I think we should pass the expr env explicitly as a map, e.g.

```golang
execute, err := shouldExecuteHook(hook.Expression, map[string]interface{} {
"step": map[string]interface{} {
  "status": ...
}
})
```